### PR TITLE
RavenDB-18149 Filter & Sort on Certificates view

### PR DESCRIPTION
**Pull request by Danielle9897**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-18149

### Additional description
Add multiple Filter & Sort options to the certs view

### Type of change
- New feature

### How risky is the change?
- Low 

### Backward compatibility
- Non breaking change

### Is it platform specific issue?
- No

### Documentation update
- This change requires a documentation update. 

### Testing 
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?
- No

### Server work needed
- RavenDB-18519

### UI work needed
RavenDB-18518

### CODE REVIEW

File: src/Raven.Client/ServerWide/Operations/Certificates/CertificateDefinition.cs [hunk 1]
```diff
@@ -46,6 +47,7 @@ public DynamicJsonValue ToJson()
                 [nameof(SecurityClearance)] = SecurityClearance,
                 [nameof(Permissions)] = permissions,
                 [nameof(NotAfter)] = NotAfter,
+                [nameof(NotBefore)] = NotBefore, // TODO - older certs will have null - see RavenDB-18519
```
Comment (ml054) [inline src/Raven.Client/ServerWide/Operations/Certificates/CertificateDefinition.cs]: so what's todo?

File: src/Raven.Studio/typescript/common/generalUtils.ts [hunk 2]
```diff
@@ -231,6 +231,15 @@ class genUtils {
 
         return input;
     }
+    
+    static sortUTC(timeStampA: string, timeStampB: string, mode: sortMode = "asc"): number {
```
Comment (ml054) [inline src/Raven.Studio/typescript/common/generalUtils.ts]: can we put extra column and sort using numbers? Parsing dates on every comparison is not good.

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts [hunk 3]
```diff
@@ -27,6 +27,9 @@ interface unifiedCertificateDefinitionWithCache extends unifiedCertificateDefini
     expirationClass: string;
     expirationText: string;
     expirationIcon: string;
+    validFromClass: string;
+    validFromText: string;
+    validFromIcon: string;
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/certificates.ts:34]: put not before, not after here as number (moment.utc(date).getDate().getTime()) and use for compare function

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts [hunk 4]
```diff
@@ -126,26 +162,57 @@ class certificates extends viewModelBase {
...
         
-        this.certificates().forEach(certificate => {
+        this.certificates().forEach((certificate: unifiedCertificateDefinitionWithCache) => {
+            
             const nameMatch = !certificate || certificate.Name.toLocaleLowerCase().includes(filter);
-            const clearanceMatch = !clearance || certificate.SecurityClearance === clearance;
             const thumbprintMatch = !certificate || _.some(certificate.Thumbprints, x => x.toLocaleLowerCase().includes(filter));
-            certificate.Visible((nameMatch || thumbprintMatch) && clearanceMatch);
+
+            const validMatch = !certificate || (this.showValidCertificates() && !certificate.expirationClass);
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/certificates.ts]: those this logic to function and return if there is no match. No need to calculate all those things

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts [hunk 5]
```diff
@@ -192,6 +259,61 @@ class certificates extends viewModelBase {
             const certs = this.certificates();
             return _.some(certs, x => x.SecurityClearance === "ClusterNode");
         });
+
+        this.canAddPermission = ko.pureComputed(() => {
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/certificates.ts]: don't block

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts [hunk 6]
```diff
@@ -192,6 +259,61 @@ class certificates extends viewModelBase {
...
+            const databaseToAdd = this.inputDatabase();
+            return databaseToAdd && !this.databasesToShow().find(x => x === databaseToAdd);
+        });
+
+        this.allDatabasesSelected = ko.pureComputed(() => !this.databasesToShow().length);
+
+        this.hasAnyFilters = ko.pureComputed(() => !!this.filterAndSortDescription() || !!this.nameFilter());
+        
+        this.filterAndSortDescription = ko.pureComputed(() => {
+            const adminPart = this.showAdminCertificates() ? "Admin clearance, " : "";
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/certificates.ts]: to array

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts [hunk 7]
```diff
@@ -192,6 +259,61 @@ class certificates extends viewModelBase {
...
+            }
+
+            const hasAllClearance = adminPart && operatorPart && userPart;
+            const clearancePart = hasAllClearance ? "" : `<strong>Clearance</strong>: ${adminPart}${operatorPart}${userPart}`.slice(0, -2);
+            
+            const hasAllStates = validPart && aboutToExpirePart && expiredPart;
+            const statePart = hasAllStates ? "" : `<strong>State</strong>: ${validPart}${aboutToExpirePart}${expiredPart}`.slice(0, -2);
+            
+            const databases = this.databasesToShow();
+            const databasesNames = databases.map(db => ` ${db}`);
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/certificates.ts]: databases.join(" ")

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts [hunk 8]
```diff
@@ -192,6 +259,61 @@ class certificates extends viewModelBase {
...
+            const databasesPart = databases.length ? `<strong>Databases</strong>: ${databasesNames}` : "";
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/certificates.ts]: escape html

File: src/Raven.Studio/wwwroot/App/views/manage/certificates.html [hunk 9]
```diff
@@ -176,6 +217,11 @@ <h3 class="no-margin" data-bind="text: Name || '<empty name>'"></h3>
                                         <h5><i class="icon-clearance"></i> SECURITY CLEARANCE</h5>
                                         <strong data-bind="text: $root.clearanceLabelFor(SecurityClearance)"></strong>
                                     </div>
+                                    <div class="valid-from">
+                                        <h5><i data-bind="attr: { class: validFromIcon }"></i> VAlID FROM</h5>
```
Comment (ml054) [inline src/Raven.Studio/wwwroot/App/views/manage/certificates.html]: VALID

Comment (Danielle9897) [reply src/Raven.Studio/wwwroot/App/views/manage/certificates.html]: done

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts (hunk 8 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/certificates.ts]: done

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts (hunk 7 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/certificates.ts]: done

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts (hunk 6 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/certificates.ts]: done

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts (hunk 5 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/certificates.ts]: done

File: src/Raven.Client/ServerWide/Operations/Certificates/CertificateDefinition.cs (hunk 1 above)
Comment (Danielle9897) [reply src/Raven.Client/ServerWide/Operations/Certificates/CertificateDefinition.cs]: removed comment

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts (hunk 4 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/certificates.ts]: done

File: src/Raven.Studio/typescript/common/generalUtils.ts (hunk 2 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/common/generalUtils.ts]: done - removed this method

File: src/Raven.Studio/typescript/viewmodels/manage/certificates.ts (hunk 3 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/certificates.ts:34]: done - used `.unix()` - similar to usage in: 
https://github.com/ravendb/ravendb/blob/887b42ba1b4c13bfa88102b241915dbab7f9c0eb/src/Raven.Studio/typescript/common/notifications/notificationCenter.ts#L179

**Review by ml054 — APPROVED** [review]

**Pull request merged**